x-pack/plugins/enterprise_search/public/applications/shared/role_mapping/users_table.tsx

```diff
@@ -112,11 +116,18 @@ export const UsersTable: React.FC<Props> = ({
     {
       field: 'accessItems',
       name: accessItemKey === 'groups' ? GROUPS_LABEL : ENGINES_LABEL,
-      render: (_, { accessItems }: SharedUser) => {
+      render: (_, { accessItems, accessAll }: SharedUser) => {
+        const isAppSearch = accessItemKey === 'engines';
         // Design calls for showing the first 2 items followed by a +x after those 2.
         // ['foo', 'bar', 'baz'] would display as: "foo, bar + 1"
         const numItems = accessItems.length;
-        if (numItems === 0) return <span data-test-subj="AllItems">{ALL_LABEL}</span>;
+        if (numItems === 0) {
+          // There is a possibility to add users without setting an access. We should not show All in that case
+          if (isAppSearch && !accessAll) {
+            return <span data-test-subj="NoItems">-</span>;
+          }
+          return <span data-test-subj="AllItems">{ALL_LABEL}</span>;
+        }
```

Comment:
The comment on line 121 (`// Design calls...`) is now out of place. After looking at this a bit, I wonder if we couldn't simplify it some and extract the logic to get the content to a function under the columns array:

```JavaScript
 const columns: Array<EuiBasicTableColumn<SharedUser>> = [
  ...
  {
    field: 'accessItems',
    name: accessItemKey === 'groups' ? GROUPS_LABEL : ENGINES_LABEL,
    render: (_, user: SharedUser) => (
      <span data-test-subj="AccessItems">{getAccessItemsContent(user)}</span>
    ),
  }
  ...
];
  
const getAccessItemsContent = ({ accessItems, accessAll }: SharedUser): string => {
  const isAppSearch = accessItemKey === 'engines';
  const numItems = accessItems.length;

  if (numItems === 0) {
    return isAppSearch && !accessAll ? '-' : ALL_LABEL;
  }

  // Design calls for showing the first 2 items followed by a +x after those 2.
  // ['foo', 'bar', 'baz'] would display as: "foo, bar + 1"
  const additionalItems = numItems > 2 ? ` + ${numItems - 2}` : '';
  const names = accessItems.map((item) => item.name);

  return names.slice(0, 2).join(', ') + additionalItems;
};
```
Then in your test, instead of asserting on the existence of `NoItems` and `AllItems`, you just assery on the content of `AccessItems`:

```JavaScript
  it('handles access items display for all items', () => {
    const userWithAllItems = {
      ...asSingleUserRoleMapping,
      roleMapping: {
        ...asRoleMapping,
        engines: [],
      },
    };
    const wrapper = mount(
      <UsersTable {...props} accessItemKey="engines" singleUserRoleMappings={[userWithAllItems]} />
    );

    expect(wrapper.find('[data-test-subj="AccessItems"]').prop('children')).toEqual('All');
  });

  it('handles access items display for no items are available.', () => {
    const userWithAllItems = {
      ...asSingleUserRoleMapping,
      roleMapping: {
        ...asRoleMapping,
        accessAllEngines: false,
        engines: [],
      },
    };
    const wrapper = mount(
      <UsersTable {...props} accessItemKey="engines" singleUserRoleMappings={[userWithAllItems]} />
    );

    expect(wrapper.find('[data-test-subj="AccessItems"]').prop('children')).toEqual('-');
  });
```

Reply:
Looks much more cleaner indeed. Updated the PR.